src/SonarQube.Client/Services/RequestFactory.cs

```diff
@@ -0,0 +1,98 @@
+using System.Collections.Generic;
+using System.Linq;
+
+namespace SonarQube.Client
+{
+    public class RequestFactory
+    {
+        private readonly Dictionary<Type, SortedList<Version, Func<object>>> requestMappings =
+            new Dictionary<Type, SortedList<Version, Func<object>>>();
+
+        /// <summary>
+        /// Registers a simple request factory for the specified version of SonarQube. It works only for types that
+        /// have parameterless constructor.
+        /// </summary>
+        /// <typeparam name="TRequest">The type of the request interface to create implementation for.</typeparam>
+        /// <typeparam name="TRequestImpl">The type implementing TRequest that has a parameterless constructor.</typeparam>
+        /// <param name="version">The version of SonarQube which first implements the request.</param>
+        /// <returns>Returns this.</returns>
+        public RequestFactory RegisterRequest<TRequest, TRequestImpl>(string version)
+            where TRequest : IRequest
+            where TRequestImpl : TRequest, new()
+        {
+            return RegisterRequest<TRequest, TRequestImpl>(version, () => new TRequestImpl());
+        }
+
+        /// <summary>
+        /// Registers a request factory for the specified version of SonarQube.
+        /// </summary>
+        /// <typeparam name="TRequest">The type of the request interface to create implementation for.</typeparam>
+        /// <typeparam name="TRequestImpl">The type implementing TRequest.</typeparam>
+        /// <param name="version">The version of SonarQube which first implements the request.</param>
+        /// <param name="factory">Factory function to create new instances of TRequestImpl.</param>
+        /// <returns>Returns this.</returns>
+        public RequestFactory RegisterRequest<TRequest, TRequestImpl>(string version, Func<TRequestImpl> factory)
+            where TRequest : IRequest
+            where TRequestImpl : TRequest
+        {
+            SortedList<Version, Func<object>> map;
+            if (!requestMappings.TryGetValue(typeof(TRequest), out map))
+            {
+                map = new SortedList<Version, Func<object>>();
+                requestMappings[typeof(TRequest)] = map;
+            }
+            map[Version.Parse(version)] = () => factory();
+            return this;
+        }
+
+        /// <summary>
+        /// Creates a new TRequest implementation for the specified SonarQube version.
+        /// </summary>
+        /// <typeparam name="TRequest">The type of the request implementation to create.</typeparam>
+        /// <param name="version">
+        /// SonarQube version to return a request implementation for. The default value returns the
+        /// latest registered implementation.QueryStringSerializer
+        /// </param>
+        /// <returns>New TRequest implementation for the specified SonarQube version.</returns>
+        public TRequest Create<TRequest>(Version version = null)
```

Comment:
Is this method ever called with Version = null?
If not, it would be better to remove it until it's needed: it's just something else to test.
(Removing it would also mean I don't have to remember why there is a design guideline about why it is better to have overloads instead of optional parameters...)